### PR TITLE
created specification builder, providers, provider manager

### PR DESCRIPTION
**Pull request by suprun-anna**

### CODE REVIEW

File: src/main/java/mate/academy/bookstore/repository/BookSpecificationProviderManager.java [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+public class BookSpecificationProviderManager implements SpecificationProviderManager<Book> {
+    private final List<SpecificationProvider<Book>> bookSpecificationProviders;
+
+    @Override
+    public SpecificationProvider<Book> getSpecificationProvider(String key) {
+        System.out.println(bookSpecificationProviders);
+        return bookSpecificationProviders.stream()
+                .filter(p -> p.getKey().equals(key))
+                .findFirst()
+                .orElseThrow(() -> new RuntimeException("Can't find correct "
```
Comment (okuzan) [inline src/main/java/mate/academy/bookstore/repository/BookSpecificationProviderManager.java]: better to throw more specific exception where possible, let's go for NoSuchElement here for example

**Review by okuzan — CHANGES_REQUESTED** [review]
Well done!

File: src/main/java/mate/academy/bookstore/repository/BookSpecificationBuilder.java [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+        if (searchParams.titles() != null && searchParams.titles().length > 0) {
+            bookSpecification = bookSpecification.and(
+                    bookSpecificationProviderManager.getSpecificationProvider(TITLE_KEY)
+                            .getSpecification(searchParams.titles()));
+        }
+
+        if (searchParams.isbns() != null && searchParams.isbns().length > 0) {
+            bookSpecification = bookSpecification.and(
+                    bookSpecificationProviderManager.getSpecificationProvider(ISBN_KEY)
+                            .getSpecification(searchParams.isbns()));
```
Comment (MentorMA) [inline src/main/java/mate/academy/bookstore/repository/BookSpecificationBuilder.java]: Duplication code, divide logic cause you have too big method

**Review by MentorMA — CHANGES_REQUESTED** [review]

File: src/main/java/mate/academy/bookstore/repository/BookSpecificationProviderManager.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+@RequiredArgsConstructor
+@Component
+public class BookSpecificationProviderManager implements SpecificationProviderManager<Book> {
+    private final List<SpecificationProvider<Book>> bookSpecificationProviders;
+
+    @Override
+    public SpecificationProvider<Book> getSpecificationProvider(String key) {
+        System.out.println(bookSpecificationProviders);
+        return bookSpecificationProviders.stream()
+                .filter(p -> p.getKey().equals(key))
```
Comment (okuzan) [inline src/main/java/mate/academy/bookstore/repository/BookSpecificationProviderManager.java:18]: change p to provider, uninformative naming is a bad idea even for stream chains

**Review by okuzan — APPROVED** [review]

**Pull request merged**